### PR TITLE
Intuitive manual position control by mapping stick input to acceleration

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by this pull request**
Final follow-up to https://github.com/PX4/Firmware/pull/12072

**Describe your solution**
The latest implementation of the stick mapping.

**Test data / coverage**
It went through 3 iterations in between and got a lot of testing on small and large multicopters and also different VTOLs.

**Additional context**
Original pr: https://github.com/PX4/Firmware/pull/12072
Split out refactoring: https://github.com/PX4/Firmware/pull/13247, https://github.com/PX4/Firmware/pull/13248, https://github.com/PX4/Firmware/pull/13262, https://github.com/PX4/Firmware/pull/13347, https://github.com/PX4/Firmware/pull/13701, https://github.com/PX4/Firmware/pull/14017
Split out acceleration based control: https://github.com/PX4/Firmware/pull/14212

### CODE REVIEW

**Review by jkflying — DISMISSED** [review]
One tiny fix, otherwise LGTM

File: src/lib/slew_rate/SlewRate.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+private:
+	Type _slew_rate{}; ///< maximum rate of change for the value
+	Type _value{}; ///< state to keep last value of the slew rate
+
+	float constrain(float value, float min, float max)
+	{
+		return math::constrain(value, min, max);
+	}
+
+	matrix::Vector3f constrain(const matrix::Vector3f &value, const matrix::Vector3f &min, const matrix::Vector3f &max)
```
Comment (dagar) [inline src/lib/slew_rate/SlewRate.hpp:91]: Might make sense in mathlib in Matrix?

File: src/lib/slew_rate/SlewRate.hpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		return _value;
+	}
+
+private:
+	Type _slew_rate{}; ///< maximum rate of change for the value
+	Type _value{}; ///< state to keep last value of the slew rate
+
+	float constrain(float value, float min, float max)
+	{
+		return math::constrain(value, min, max);
```
Comment (dagar) [inline src/lib/slew_rate/SlewRate.hpp:88]: Why does this exist?

File: src/modules/mc_pos_control/mc_pos_control_params.c [hunk 3]
```diff
@@ -727,13 +730,14 @@ PARAM_DEFINE_FLOAT(MPC_TKO_RAMP_T, 3.0f);
  * 1 Smooth position control with maximum acceleration and jerk limits based on slew-rates.
  * 3 Smooth position control with maximum acceleration and jerk limits based on
  *   jerk optimized trajectory generator (different algorithm than 1).
+ * 4 Smooth position control where sticks map to acceleration and there's a virtual brake drag
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/mc_pos_control/mc_pos_control_params.c:733]: Thanks for the reminder. I forgot this line.

File: src/lib/slew_rate/SlewRate.hpp (hunk 1 above)
Comment (MaEtUgR) [reply src/lib/slew_rate/SlewRate.hpp:91]: Yes, would be better there 👍

File: src/lib/slew_rate/SlewRate.hpp (hunk 2 above)
Comment (MaEtUgR) [reply src/lib/slew_rate/SlewRate.hpp:88]: To have one implementation for float and one for Vector3f. I'll make it supported by the matrix library and this local implementation can go away.

File: src/modules/mc_pos_control/mc_pos_control_params.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/mc_pos_control/mc_pos_control_params.c:733]: Added ✔️

File: src/lib/slew_rate/SlewRate.hpp (hunk 1 above)
Comment (MaEtUgR) [reply src/lib/slew_rate/SlewRate.hpp:91]: There actually already is an implementation: https://github.com/PX4/Matrix/blob/ce6b10b99af722a4e653387d40019a8428d2a54f/matrix/Matrix.hpp#L651-L660
I guess there was none when I started out initially...

Comment (MaEtUgR) [reply src/lib/slew_rate/SlewRate.hpp]: Do you prefer this solution?
https://github.com/PX4/Firmware/pull/16052/commits/f6b74ee92840bd1dfa6be66411b0e03d679a4bfb

**Comment by nrogelio** [comment]
Vehicle: S500 with Pixhawk 4

It was a windy day, but the vehicle was responsible and flew well.

I noticed then I release the sticks in Stabilize it maintained it's position like in position mode.

Logs:
https://review.px4.io/plot_app?log=91820a9d-1170-4d8c-9921-13ecbab9b348
https://review.px4.io/plot_app?log=ca7a139f-ae04-47c0-8954-fad4dcd9f9e4
https://review.px4.io/plot_app?log=bcb5a3b5-691b-48d3-8770-3caab1746580
https://review.px4.io/plot_app?log=798e509b-bb5a-44a1-97e8-0399413d5b9c

**Comment by MaEtUgR** [comment]
@nrogelio Nice. Thanks so much for testing and the feedback!

> I noticed then I release the sticks in Stabilize it maintained it's position like in position mode.

Are you sure it wasn't in position mode? That's the point of the new position mode to feel more like altitude mode but have the advantages of position mode.

**Comment by MaEtUgR** [comment]
I checked in simulation and altitude and stabilized mode are still behaving like before.
I also checked your logs and the setpoint generation in the different modes looks like expected.
I'll merge this when CI is happy.

**Comment by nrogelio** [comment]
> Are you sure it wasn't in position mode? That's the point of the new position mode to feel more like altitude mode but have the advantages of position mode. I'll still double check.

I was in Stabilize, but you're right, it behaved more like Altitude, maybe the angle was giving me the impression that it wasn't displacing.

Check https://review.px4.io/3d?log=bcb5a3b5-691b-48d3-8770-3caab1746580 at 8:18

**Comment by MaEtUgR** [comment]
> maybe the angle was giving me the impression that it wasn't displacing.

Yeah, angle goes to zero and it's coasting with air resistance but without any velocity setpoint like it's expected in stabilized. It's probably by accident so slow. Nothing changed with stabilized mode in this pr and the modes are still called correctly.

image

**Comment by MaEtUgR** [comment]
I removed the old old position mode smoothing to save flash, 1082 lines 👋 
Realized on the way that orbit mode was still using the old smoothing for the altitude 🔍

**Pull request merged**